File: x402/mechanisms/casper/exact/facilitator/scheme.go
package facilitator

import (
	"context"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"math/big"
	"strconv"
	"strings"
	"time"

	eip712 "github.com/casper-ecosystem/casper-eip-712/go"
	"github.com/make-software/casper-go-sdk/v2/casper"
	"github.com/make-software/casper-go-sdk/v2/types"
	"github.com/make-software/casper-go-sdk/v2/types/clvalue"
	"github.com/make-software/casper-go-sdk/v2/types/clvalue/cltype"
	"github.com/make-software/casper-go-sdk/v2/types/key"
	"github.com/make-software/casper-go-sdk/v2/types/keypair"

	casperMechanism "casper_x402_facilitator/x402/mechanisms/casper"

	x402 "github.com/x402-foundation/x402/go"
	x402types "github.com/x402-foundation/x402/go/types"
)

const defaultPaymentMotes uint64 = 7_000_000_000

var transferWithAuthorizationTypes = eip712.TypeDefinitions{
	"TransferWithAuthorization": {
		{Name: "from", Type: "address"},
		{Name: "to", Type: "address"},
		{Name: "value", Type: "uint256"},
		{Name: "validAfter", Type: "uint256"},
		{Name: "validBefore", Type: "uint256"},
		{Name: "nonce", Type: "bytes32"},
	},
}

type ExactCasperSchemeConfig struct {
	LimitedPaymentMotes *uint64
}

type ExactCasperScheme struct {
	signer casperMechanism.FacilitatorCasperSigner
	config *ExactCasperSchemeConfig
}

func NewExactCasperScheme(signer casperMechanism.FacilitatorCasperSigner, config *ExactCasperSchemeConfig) *ExactCasperScheme {
	if config == nil {
		config = &ExactCasperSchemeConfig{}
	}

	return &ExactCasperScheme{signer: signer, config: config}
}

func (f *ExactCasperScheme) Scheme() string {
	return casperMechanism.SchemeExact
}

func (f *ExactCasperScheme) CaipFamily() string {
	return "casper:*"
}

func (f *ExactCasperScheme) GetExtra(network x402.Network) map[string]interface{} {
	feePayer := ""
	addresses := f.signer.GetAddresses(context.Background(), string(network))
	if len(addresses) > 0 {
		feePayer = addresses[0]
	}

	return map[string]interface{}{
		"feePayer": feePayer,
	}
}

func (f *ExactCasperScheme) GetSigners(network x402.Network) []string {
	return f.signer.GetAddresses(context.Background(), string(network))
}

func (f *ExactCasperScheme) Verify(
	ctx context.Context,
	payload x402types.PaymentPayload,
	requirements x402types.PaymentRequirements,
	_ *x402.FacilitatorContext,
) (*x402.VerifyResponse, error) {
	if payload.Accepted.Scheme != casperMechanism.SchemeExact {
		return nil, x402.NewVerifyError(ErrUnsupportedScheme, "", fmt.Sprintf("invalid scheme: %s", payload.Accepted.Scheme))
	}

	if payload.Accepted.Network != requirements.Network {
		return nil, x402.NewVerifyError(ErrNetworkMismatch, "", fmt.Sprintf("network mismatch: payload=%s requirements=%s", payload.Accepted.Network, requirements.Network))
	}

	p, err := payloadFromMap(payload.Payload)
	if err != nil {
		return nil, x402.NewVerifyError(ErrMalformedPayload, "", err.Error())
	}

	if p.Authorization.To != requirements.PayTo {
		return nil, x402.NewVerifyError(ErrPayToMismatch, "", fmt.Sprintf("authorization.to=%s requirements.payTo=%s", p.Authorization.To, requirements.PayTo))
	}

	if p.Authorization.Value != requirements.Amount {
		return nil, x402.NewVerifyError(ErrAmountMismatch, "", fmt.Sprintf("authorization.value=%s requirements.amount=%s", p.Authorization.Value, requirements.Amount))
	}

	if !casperMechanism.IsValidAddress(requirements.PayTo) || !casperMechanism.IsValidAddress(p.Authorization.To) {
		return nil, x402.NewVerifyError(ErrInvalidPayTo, "", "invalid payTo account-hash")
	}

	if requirements.Amount == "" || requirements.Amount == "0" || p.Authorization.Value == "" || p.Authorization.Value == "0" {
		return nil, x402.NewVerifyError(ErrInvalidAmount, "", "amount must be non-zero")
	}

	if !casperMechanism.IsValidContractPackageHash(requirements.Asset) {
		return nil, x402.NewVerifyError(ErrInvalidAsset, "", requirements.Asset)
	}

	validAfter, err := strconv.ParseInt(p.Authorization.ValidAfter, 10, 64)
	if err != nil {
		return nil, x402.NewVerifyError(ErrMalformedPayload, "", "invalid validAfter")
	}
	validBefore, err := strconv.ParseInt(p.Authorization.ValidBefore, 10, 64)
	if err != nil {
		return nil, x402.NewVerifyError(ErrMalformedPayload, "", "invalid validBefore")
	}

	now := time.Now().Unix()
	if validAfter > now {
		return nil, x402.NewVerifyError(ErrNotYetValid, "", fmt.Sprintf("validAfter=%d > now=%d", validAfter, now))
	}
	if now > validBefore {
		return nil, x402.NewVerifyError(ErrPayloadExpired, "", fmt.Sprintf("validBefore=%d < now=%d", validBefore, now))
	}
	if validBefore-now < 6 {
		return nil, x402.NewVerifyError(ErrInsufficientTime, "", "less than 6s to settle")
	}

	// verify we have a signer for the network in requirements
	cfg, err := f.signer.GetNetworkConfig(ctx, requirements.Network)
	if err != nil {
		return nil, x402.NewVerifyError(ErrNetworkMismatch, "", err.Error())
	}

	name, ok := requirements.Extra["name"].(string)
	if !ok || name == "" {
		return nil, x402.NewVerifyError(ErrMissingTokenName, "", "")
	}
	version, ok := requirements.Extra["version"].(string)
	if !ok || version == "" {
		return nil, x402.NewVerifyError(ErrMissingTokenVersion, "", "")
	}

	contractPackageHash, err := casperMechanism.DecodeContractPackageHash(requirements.Asset)
	if err != nil {
		return nil, x402.NewVerifyError(ErrInvalidAsset, "", err.Error())
	}

	domain := eip712.BuildDomain(name, version, cfg.ChainName, contractPackageHash)

	fromAddr, err := eip712.NewAddressFromHex("0x" + p.Authorization.From)
	if err != nil {
		return nil, x402.NewVerifyError(ErrMalformedPayload, "", fmt.Sprintf("invalid from address: %v", err))
	}
	toAddr, err := eip712.NewAddressFromHex("0x" + p.Authorization.To)
	if err != nil {
		return nil, x402.NewVerifyError(ErrMalformedPayload, "", fmt.Sprintf("invalid to address: %v", err))
	}

	value, ok := new(big.Int).SetString(p.Authorization.Value, 10)
	if !ok {
		return nil, x402.NewVerifyError(ErrMalformedPayload, "", "invalid value")
	}

	nonceRaw, err := hex.DecodeString(p.Authorization.Nonce)
	if err != nil || len(nonceRaw) != 32 {
		return nil, x402.NewVerifyError(ErrMalformedPayload, "", "invalid nonce")
	}
	var nonceBytes [32]byte
	copy(nonceBytes[:], nonceRaw)

	message := map[string]interface{}{
		"from":        fromAddr,
		"to":          toAddr,
		"value":       value,
		"validAfter":  big.NewInt(validAfter),
		"validBefore": big.NewInt(validBefore),
		"nonce":       nonceBytes,
	}

	digest, err := eip712.HashTypedData(domain, transferWithAuthorizationTypes, "TransferWithAuthorization", message, &eip712.TypedDataOptions{DomainTypes: eip712.CasperDomainTypes})
	if err != nil {
		return nil, x402.NewVerifyError(ErrFailedToHash, "", err.Error())
	}

	sigBytes, err := hex.DecodeString(p.Signature)
	if err != nil || len(sigBytes) != 65 {
		return nil, x402.NewVerifyError(ErrInvalidSignature, "", "signature must be 65 bytes hex")
	}
	var sig65 [65]byte
	copy(sig65[:], sigBytes)

	verified, err := f.signer.VerifyEIP712Signature(digest, sig65, p.PublicKey)
	if err != nil {
		return nil, x402.NewVerifyError(ErrInvalidSignature, "", err.Error())
	}
	if !verified {
		return nil, x402.NewVerifyError(ErrInvalidSignature, "", "signature verification failed")
	}

	return &x402.VerifyResponse{IsValid: true, Payer: p.Authorization.From}, nil
}

func (f *ExactCasperScheme) paymentMotes() uint64 {
	if f.config.LimitedPaymentMotes != nil {
		return *f.config.LimitedPaymentMotes
	}

	return defaultPaymentMotes
}

func (f *ExactCasperScheme) Settle(
	ctx context.Context,
	payload x402types.PaymentPayload,
	requirements x402types.PaymentRequirements,
	fctx *x402.FacilitatorContext,
) (*x402.SettleResponse, error) {
	network := x402.Network(requirements.Network)

	verifyResp, err := f.Verify(ctx, payload, requirements, fctx)
	if err != nil {
		ve := &x402.VerifyError{}
		if errors.As(err, &ve) {
			return nil, x402.NewSettleError(ErrVerificationFailed, ve.Payer, network, "", ve.InvalidMessage)
		}
		return nil, x402.NewSettleError(ErrVerificationFailed, "", network, "", err.Error())
	}

	p, err := payloadFromMap(payload.Payload)
	if err != nil {
		return nil, x402.NewSettleError(ErrMalformedPayload, verifyResp.Payer, network, "", err.Error())
	}

	args := types.Args{}

	fromKey, err := key.NewKey("account-hash-" + p.Authorization.From[2:])
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", err.Error())
	}
	toKey, err := key.NewKey("account-hash-" + p.Authorization.To[2:])
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", err.Error())
	}
	amountInt, ok := new(big.Int).SetString(p.Authorization.Value, 10)
	if !ok {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", "invalid amount")
	}
	validAfterInt, err := strconv.ParseUint(p.Authorization.ValidAfter, 10, 64)
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", "invalid validAfter")
	}
	validBeforeInt, err := strconv.ParseUint(p.Authorization.ValidBefore, 10, 64)
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", "invalid validBefore")
	}
	nonceBytes, err := hex.DecodeString(p.Authorization.Nonce)
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", err.Error())
	}
	sigBytes, err := hex.DecodeString(p.Signature)
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", err.Error())
	}
	pubKey, err := keypair.NewPublicKey(p.PublicKey)
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", err.Error())
	}

	args.AddArgument("from", clvalue.NewCLKey(fromKey)).
		AddArgument("to", clvalue.NewCLKey(toKey)).
		AddArgument("amount", *clvalue.NewCLUInt256(amountInt)).
		AddArgument("valid_after", *clvalue.NewCLUInt64(validAfterInt)).
		AddArgument("valid_before", *clvalue.NewCLUInt64(validBeforeInt)).
		AddArgument("nonce", clListU8WithBytes(nonceBytes)).
		AddArgument("public_key", clvalue.NewCLPublicKey(pubKey)).
		AddArgument("signature", clListU8WithBytes(sigBytes))

	packageHash, err := casper.NewHash(requirements.Asset)
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", err.Error())
	}
	entryPoint := "transfer_with_authorization"

	facilitatorPubKeyHex, err := f.signer.GetPublicKeyHex(ctx, requirements.Network)
	if err != nil || facilitatorPubKeyHex == "" {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", "no facilitator public key")
	}
	facilitatorPubKey, err := keypair.NewPublicKey(facilitatorPubKeyHex)
	if err != nil {
		return nil, x402.NewSettleError(ErrBuildDeployFailed, verifyResp.Payer, network, "", err.Error())
	}

	// verify we have a signer for this network before proceeding
	networkCfg, err := f.signer.GetNetworkConfig(ctx, requirements.Network)
	if err != nil {
		return nil, x402.NewSettleError(ErrNetworkMismatch, verifyResp.Payer, network, "", err.Error())
	}

	chainName := strings.TrimPrefix(networkCfg.ChainName, "casper:")

	v1Payload, err := types.NewTransactionV1Payload(
		types.InitiatorAddr{
			PublicKey: &facilitatorPubKey,
		},
		types.Timestamp(time.Now().UTC()),
		900000000000,
		chainName,
		types.PricingMode{
			Limited: &types.LimitedMode{
				GasPriceTolerance: 1,
				StandardPayment:   true,
				PaymentAmount:     f.paymentMotes(),
			},
		},
		types.NewNamedArgs(&args),
		types.TransactionTarget{
			Stored: &types.StoredTarget{
				ID: types.TransactionInvocationTarget{
					ByPackageHash: &types.ByPackageHashInvocationTarget{
						Addr:    packageHash,
						Version: nil,
					},
				},
				Runtime: types.NewVmCasperV1TransactionRuntime(),
			},
		},
		types.TransactionEntryPoint{
			Custom: &entryPoint,
		},
		types.TransactionScheduling{
			Standard: &struct{}{},
		},
	)

	if err != nil {
		return nil, x402.NewSettleError(ErrSignDeployFailed, verifyResp.Payer, network, "", err.Error())
	}

	transaction, err := types.MakeTransactionV1(v1Payload)
	if err != nil {
		return nil, x402.NewSettleError(ErrSignDeployFailed, verifyResp.Payer, network, "", err.Error())
	}

	if err := f.signer.SignTransaction(transaction, requirements.Network); err != nil {
		return nil, x402.NewSettleError(ErrSignDeployFailed, verifyResp.Payer, network, "", err.Error())
	}

	_, err = f.signer.PutTransaction(ctx, requirements.Network, *transaction)
	if err != nil {
		return nil, x402.NewSettleError(ErrPutDeployFailed, verifyResp.Payer, network, transaction.Hash.String(), err.Error())
	}

	if err := f.signer.WaitForTransaction(ctx, requirements.Network, transaction.Hash.String()); err != nil {
		return nil, x402.NewSettleError(ErrWaitDeployFailed, verifyResp.Payer, network, transaction.Hash.String(), err.Error())
	}

	return &x402.SettleResponse{
		Success:     true,
		Transaction: transaction.Hash.String(),
		Network:     network,
		Payer:       verifyResp.Payer,
	}, nil
}

func payloadFromMap(data map[string]interface{}) (*casperMechanism.ExactCasperPayload, error) {
	if data == nil {
		return nil, fmt.Errorf("payload is nil")
	}

	jsonBytes, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload data: %w", err)
	}

	var payload casperMechanism.ExactCasperPayload
	if err := json.Unmarshal(jsonBytes, &payload); err != nil {
		return nil, fmt.Errorf("failed to unmarshal payload: %w", err)
	}

	if payload.Signature == "" || payload.PublicKey == "" {
		return nil, fmt.Errorf("missing signature or publicKey field")
	}

	if payload.Authorization.From == "" ||
		payload.Authorization.To == "" ||
		payload.Authorization.Value == "" ||
		payload.Authorization.ValidAfter == "" ||
		payload.Authorization.ValidBefore == "" ||
		payload.Authorization.Nonce == "" {
		return nil, fmt.Errorf("missing authorization fields")
	}

	return &payload, nil
}

func clListU8WithBytes(bytes []byte) clvalue.CLValue {
	elements := make([]clvalue.CLValue, 0)
	for _, v := range bytes {
		ui8 := clvalue.UInt8(v)
		elements = append(elements, clvalue.CLValue{
			Type: cltype.UInt8,
			UI8:  &ui8,
		})
	}
	clType := cltype.NewList(cltype.UInt8)
	value := clvalue.CLValue{
		Type: clType,
		List: &clvalue.List{
			Type:     clType,
			Elements: elements,
		},
	}

	return value
}

File: js/packages/mechanisms/casper/src/exact/facilitator/scheme.ts
import {
  PaymentPayload,
  PaymentRequirements,
  SchemeNetworkFacilitator,
  FacilitatorContext,
  SettleResponse,
  VerifyResponse,
  Network,
} from "@x402/core/types";
import { buildDomain, CASPER_DOMAIN_TYPES, hashTypedData } from "@casper-ecosystem/casper-eip-712";
import casperSdk from "casper-js-sdk";
import { SCHEME_EXACT } from "../../constants";
import { FacilitatorCasperSigner } from "../../signer";
import { ExactCasperPayload } from "../../types";
import { chainNameFromNetwork, isValidAddress, isValidContractPackageHash } from "../../utils";

const DEFAULT_PAYMENT_MOTES = 7_000_000_000;

/**
 * Decode a hex string into a Uint8Array.
 *
 * @param hex - Hex string, with or without "0x" prefix.
 * @returns Decoded bytes.
 * @throws Error if the hex string is invalid.
 */
function hexToBytes(hex: string): Uint8Array {
  const cleaned = hex.startsWith("0x") || hex.startsWith("0X") ? hex.slice(2) : hex;
  if (cleaned.length % 2 !== 0) {
    throw new Error("hex string must have an even number of characters");
  }
  const bytes = new Uint8Array(cleaned.length / 2);
  for (let i = 0; i < cleaned.length; i += 2) {
    bytes[i / 2] = parseInt(cleaned.substring(i, i + 2), 16);
  }
  return bytes;
}

const transferWithAuthorizationTypes = {
  TransferWithAuthorization: [
    { name: "from", type: "address" },
    { name: "to", type: "address" },
    { name: "value", type: "uint256" },
    { name: "validAfter", type: "uint256" },
    { name: "validBefore", type: "uint256" },
    { name: "nonce", type: "bytes32" },
  ],
};

export const ErrInvalidScheme = "invalid_exact_casper_facilitator_invalid_scheme";
export const ErrNetworkMismatch = "invalid_exact_casper_facilitator_network_mismatch";
export const ErrInvalidAsset = "invalid_exact_casper_facilitator_invalid_asset";
export const ErrInvalidPayTo = "invalid_exact_casper_facilitator_invalid_payto";
export const ErrInvalidPayer = "invalid_exact_casper_facilitator_invalid_payer";
export const ErrAmountMismatch = "invalid_exact_casper_facilitator_amount_mismatch";
export const ErrPayToMismatch = "invalid_exact_casper_facilitator_payto_mismatch";
export const ErrExpired = "invalid_exact_casper_facilitator_expired";
export const ErrNotYetValid = "invalid_exact_casper_facilitator_not_yet_valid";
export const ErrInvalidSignature = "invalid_exact_casper_facilitator_invalid_signature";
export const ErrPublicKeyMismatch = "invalid_exact_casper_facilitator_publickey_mismatch";
export const ErrSettleFailed = "invalid_exact_casper_facilitator_settle_failed";
export const ErrMissingTokenName = "invalid_exact_casper_facilitator_missing_token_name";
export const ErrMissingTokenVersion = "invalid_exact_casper_facilitator_missing_token_version";
export const ErrFailedToHash = "invalid_exact_casper_facilitator_failed_to_hash";

export interface ExactCasperSchemeConfig {
  limitedPaymentMotes?: number;
}

/**
 * Casper facilitator implementation for the Exact payment scheme.
 */
export class ExactCasperScheme implements SchemeNetworkFacilitator {
  readonly scheme = SCHEME_EXACT;
  readonly caipFamily = "casper:*";

  /**
   * Creates a new ExactCasperScheme facilitator instance.
   *
   * @param signer - The Casper signer for facilitator operations
   * @param config - Optional facilitator configuration
   */
  constructor(
    private readonly signer: FacilitatorCasperSigner,
    private readonly config?: ExactCasperSchemeConfig,
  ) {}

  /**
   * Get mechanism-specific extra data for the supported kinds endpoint.
   *
   * @param network - The network identifier
   * @returns Extra data object with feePayer
   */
  getExtra(network: Network): Record<string, unknown> {
    const addresses = this.signer.getAddresses(network);
    return {
      feePayer: addresses[0] ?? "",
    };
  }

  /**
   * Get signer addresses used by this facilitator for a given network.
   *
   * @param network - The network identifier
   * @returns Array of signer addresses
   */
  getSigners(network: Network): string[] {
    return this.signer.getAddresses(network);
  }

  /**
   * Verify a Casper Exact payment payload.
   *
   * @param payload - The payment payload to verify
   * @param requirements - The payment requirements
   * @param _ - Optional facilitator context
   * @returns Promise resolving to verification response
   */
  async verify(
    payload: PaymentPayload,
    requirements: PaymentRequirements,
    _?: FacilitatorContext,
  ): Promise<VerifyResponse> {
    if (payload.accepted.scheme !== SCHEME_EXACT || requirements.scheme !== SCHEME_EXACT) {
      return { isValid: false, invalidReason: ErrInvalidScheme };
    }

    if (payload.accepted.network !== requirements.network) {
      return {
        isValid: false,
        invalidReason: ErrNetworkMismatch,
        invalidMessage: `payload=${payload.accepted.network} requirements=${requirements.network}`,
      };
    }

    let p: ExactCasperPayload;
    try {
      p = payload.payload as ExactCasperPayload;
      if (
        !p.signature ||
        !p.publicKey ||
        !p.authorization.from ||
        !p.authorization.to ||
        p.authorization.value === undefined ||
        !p.authorization.validAfter ||
        !p.authorization.validBefore ||
        !p.authorization.nonce
      ) {
        throw new Error("missing payload fields");
      }
    } catch {
      return {
        isValid: false,
        invalidReason: ErrInvalidScheme,
        invalidMessage: "malformed payload",
      };
    }

    const payer = p.authorization.from;

    if (p.authorization.to !== requirements.payTo) {
      return {
        isValid: false,
        invalidReason: ErrPayToMismatch,
        invalidMessage: `authorization.to=${p.authorization.to} requirements.payTo=${requirements.payTo}`,
        payer,
      };
    }

    if (p.authorization.value !== requirements.amount) {
      return {
        isValid: false,
        invalidReason: ErrAmountMismatch,
        invalidMessage: `authorization.value=${p.authorization.value} requirements.amount=${requirements.amount}`,
        payer,
      };
    }

    if (!isValidAddress(requirements.payTo) || !isValidAddress(p.authorization.to)) {
      return { isValid: false, invalidReason: ErrInvalidPayTo, payer };
    }

    if (!isValidAddress(payer)) {
      return { isValid: false, invalidReason: ErrInvalidPayer, payer };
    }

    if (
      requirements.amount === "" ||
      requirements.amount === "0" ||
      p.authorization.value === "" ||
      p.authorization.value === "0"
    ) {
      return {
        isValid: false,
        invalidReason: ErrAmountMismatch,
        invalidMessage: "amount must be non-zero",
        payer,
      };
    }

    if (!isValidContractPackageHash(requirements.asset)) {
      return {
        isValid: false,
        invalidReason: ErrInvalidAsset,
        invalidMessage: requirements.asset,
        payer,
      };
    }

    let validAfter: number;
    let validBefore: number;
    try {
      validAfter = parseInt(p.authorization.validAfter, 10);
      validBefore = parseInt(p.authorization.validBefore, 10);
      if (isNaN(validAfter) || isNaN(validBefore)) {
        throw new Error("invalid timestamp");
      }
    } catch {
      return {
        isValid: false,
        invalidReason: ErrInvalidScheme,
        invalidMessage: "invalid validAfter/validBefore",
        payer,
      };
    }

    const now = Math.floor(Date.now() / 1000);
    if (validAfter > now) {
      return {
        isValid: false,
        invalidReason: ErrNotYetValid,
        invalidMessage: `validAfter=${validAfter} > now=${now}`,
        payer,
      };
    }
    if (now > validBefore) {
      return {
        isValid: false,
        invalidReason: ErrExpired,
        invalidMessage: `validBefore=${validBefore} < now=${now}`,
        payer,
      };
    }
    if (validBefore - now < 6) {
      return {
        isValid: false,
        invalidReason: ErrExpired,
        invalidMessage: "less than 6s to settle",
        payer,
      };
    }

    const name = requirements.extra?.name;
    const version = requirements.extra?.version;
    if (typeof name !== "string" || name === "") {
      return { isValid: false, invalidReason: ErrMissingTokenName, payer };
    }
    if (typeof version !== "string" || version === "") {
      return { isValid: false, invalidReason: ErrMissingTokenVersion, payer };
    }

    try {
      await this.signer.getNetworkConfig(requirements.network);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return { isValid: false, invalidReason: ErrNetworkMismatch, invalidMessage: message, payer };
    }

    let digest: Uint8Array;
    try {
      const domain = buildDomain(name, version, requirements.network, "0x" + requirements.asset);

      const nonceBytes = hexToBytes(p.authorization.nonce);
      if (nonceBytes.length !== 32) {
        return {
          isValid: false,
          invalidReason: ErrInvalidSignature,
          invalidMessage: "nonce must be 32 bytes",
          payer,
        };
      }

      const message = {
        from: "0x" + payer,
        to: "0x" + p.authorization.to,
        value: BigInt(p.authorization.value),
        validAfter: BigInt(validAfter),
        validBefore: BigInt(validBefore),
        nonce: "0x" + p.authorization.nonce,
      };

      digest = hashTypedData(
        domain,
        transferWithAuthorizationTypes,
        "TransferWithAuthorization",
        message,
        {
          domainTypes: CASPER_DOMAIN_TYPES,
        },
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return { isValid: false, invalidReason: ErrFailedToHash, invalidMessage: message, payer };
    }

    let publicKey: casperSdk.PublicKey;
    try {
      publicKey = casperSdk.PublicKey.fromHex(p.publicKey);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return { isValid: false, invalidReason: ErrInvalidSignature, invalidMessage: message, payer };
    }

    if (publicKey.accountHash().toHex() !== payer.slice(2)) {
      return {
        isValid: false,
        invalidReason: ErrPublicKeyMismatch,
        invalidMessage: "public key does not match authorization.from",
        payer,
      };
    }

    let signatureBytes: Uint8Array;
    try {
      signatureBytes = hexToBytes(p.signature);
      if (signatureBytes.length !== 65) {
        return {
          isValid: false,
          invalidReason: ErrInvalidSignature,
          invalidMessage: "signature must be 65 bytes",
          payer,
        };
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return { isValid: false, invalidReason: ErrInvalidSignature, invalidMessage: message, payer };
    }

    try {
      if (!publicKey.verifySignature(digest, signatureBytes)) {
        return {
          isValid: false,
          invalidReason: ErrInvalidSignature,
          invalidMessage: "signature verification failed",
          payer,
        };
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return { isValid: false, invalidReason: ErrInvalidSignature, invalidMessage: message, payer };
    }

    return { isValid: true, payer };
  }

  /**
   * Settle a Casper Exact payment payload.
   *
   * @param payload - The payment payload to settle
   * @param requirements - The payment requirements
   * @param context - Optional facilitator context
   * @returns Promise resolving to settlement response
   */
  async settle(
    payload: PaymentPayload,
    requirements: PaymentRequirements,
    context?: FacilitatorContext,
  ): Promise<SettleResponse> {
    const network = requirements.network;

    const verifyResp = await this.verify(payload, requirements, context);
    if (!verifyResp.isValid) {
      return {
        success: false,
        errorReason: verifyResp.invalidReason,
        errorMessage: verifyResp.invalidMessage,
        payer: verifyResp.payer,
        transaction: "",
        network,
      };
    }

    const p = payload.payload as ExactCasperPayload;

    let publicKeyHex: string;
    try {
      publicKeyHex = this.signer.getPublicKeyHex(network);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        success: false,
        errorReason: ErrSettleFailed,
        errorMessage: message,
        payer: verifyResp.payer,
        transaction: "",
        network,
      };
    }

    let facilitatorPublicKey: casperSdk.PublicKey;
    try {
      facilitatorPublicKey = casperSdk.PublicKey.fromHex(publicKeyHex);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        success: false,
        errorReason: ErrSettleFailed,
        errorMessage: message,
        payer: verifyResp.payer,
        transaction: "",
        network,
      };
    }

    let networkConfig;
    try {
      networkConfig = await this.signer.getNetworkConfig(network);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        success: false,
        errorReason: ErrNetworkMismatch,
        errorMessage: message,
        payer: verifyResp.payer,
        transaction: "",
        network,
      };
    }

    try {
      const args = buildTransferWithAuthorizationArgs(p);

      const transaction = new casperSdk.ContractCallBuilder()
        .from(facilitatorPublicKey)
        .byPackageHash(requirements.asset)
        .entryPoint("transfer_with_authorization")
        .runtimeArgs(args)
        .chainName(chainNameFromNetwork(networkConfig.chainName))
        .payment(this.config?.limitedPaymentMotes ?? DEFAULT_PAYMENT_MOTES)
        .build();

      await this.signer.signTransaction(transaction, network);
      const transactionHash = await this.signer.putTransaction(network, transaction);
      await this.signer.waitForTransaction(network, transactionHash);

      return {
        success: true,
        transaction: transactionHash,
        network,
        payer: verifyResp.payer,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        success: false,
        errorReason: ErrSettleFailed,
        errorMessage: message,
        payer: verifyResp.payer,
        transaction: "",
        network,
      };
    }
  }
}

/**
 * Build runtime args for the transfer_with_authorization entry point.
 *
 * @param payload - The Exact Casper payload
 * @returns Casper Args
 */
function buildTransferWithAuthorizationArgs(payload: ExactCasperPayload) {
  const fromKey = casperSdk.Key.newKey("account-hash-" + payload.authorization.from.slice(2));
  const toKey = casperSdk.Key.newKey("account-hash-" + payload.authorization.to.slice(2));
  const signatureBytes = hexToBytes(payload.signature);
  const nonceBytes = hexToBytes(payload.authorization.nonce);
  const publicKey = casperSdk.PublicKey.fromHex(payload.publicKey);

  const args: Record<string, casperSdk.CLValue> = {
    from: casperSdk.CLValue.newCLKey(fromKey),
    to: casperSdk.CLValue.newCLKey(toKey),
    value: casperSdk.CLValue.newCLUInt256(payload.authorization.value),
    valid_after: casperSdk.CLValue.newCLUint64(parseInt(payload.authorization.validAfter, 10)),
    valid_before: casperSdk.CLValue.newCLUint64(parseInt(payload.authorization.validBefore, 10)),
    nonce: casperSdk.CLValue.newCLList(
      casperSdk.CLTypeUInt8,
      Array.from(nonceBytes).map(b => casperSdk.CLValue.newCLUint8(b)),
    ),
    public_key: casperSdk.CLValue.newCLPublicKey(publicKey),
    signature: casperSdk.CLValue.newCLList(
      casperSdk.CLTypeUInt8,
      Array.from(signatureBytes).map(b => casperSdk.CLValue.newCLUint8(b)),
    ),
  };

  return casperSdk.Args.fromMap(args);
}
